pyvcg/driver/file_smtlib.py

```diff
@@ -172,13 +172,24 @@ def visit_enumeration_declaration(self, node):
     def visit_record_declaration(self, node):
         assert isinstance(node, smt.Record_Declaration)
         self.emit_comment(node.comment)
-        self.lines.append("(declare-datatype %s ((%s" %
-                          (self.escape_name(node.sort.name),
-                           self.escape_name(node.sort.name + "__cons")))
-        for name, sort in node.sort.components.items():
-            self.lines.append("  (%s %s)" % (self.escape_name(name),
-                                             sort.walk(self)))
-        self.lines[-1] += ")))"
+
+        sort_name = self.escape_name(node.sort.name)
+        self.lines.append(f"(declare-datatype {sort_name} (")
+
+        if node.is_recursive():
```

Comment:
I also think this will only work for direct top-level recursion.

However a record A with field type (also a record) B that then references A would not work if I read this right.

The SMTLIB construct to use is `declare-datatypes` (see p64 in SMTLIB 2.7).

I think we should perhaps add a Recursive_Record structure that allows you to do that; otherwise need need to do analysis of the tree and figure out which records to group together.